services/wallet/history/balance.go

```diff
@@ -0,0 +1,301 @@
+				Timestamp: uint64(cached[0].timestamp),
+			}, blockNo, nil
+		}
+
+		// otherwise try fetch any to get the timestamp info
+		outEntry, _, err = b.db.getFirst(source.ChainID(), blockNo)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+	var timestamp int64
+	if outEntry != nil {
+		timestamp = outEntry.timestamp
+	} else {
+		header, err := source.HeaderByNumber(ctx, blockNo)
+		if err != nil {
+			return nil, nil, err
+		}
+		blockNo = new(big.Int).Set(header.Number)
+		timestamp = int64(header.Time)
+	}
+
+	currentBalance, err := source.BalanceAt(ctx, address, blockNo)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	entry := entry{
+		chainID:   source.ChainID(),
+		address:   address,
+		currency:  source.Currency(),
+		block:     new(big.Int).Set(blockNo),
+		balance:   currentBalance,
+		timestamp: timestamp,
+	}
+	err = b.db.add(&entry, bitset)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var dataPoint DataPoint
+	dataPoint.Value = (*hexutil.Big)(currentBalance)
+	dataPoint.Timestamp = uint64(timestamp)
+	return &dataPoint, blockNo, nil
+}
+
+func (b *Balance) update(ctx context.Context, source DataSource, address common.Address, endingTimestamp int64, timeInterval TimeInterval) error {
+	startTimestamp := int64(0)
+	fetchTimestamp := int64(0)
+	if timeInterval != BalanceHistoryAllTime {
+		// Ensure we always get the complete range by fetching the next block also
+		startTimestamp = endingTimestamp - int64(timeIntervalDuration[timeInterval].Seconds())
+		fetchTimestamp = startTimestamp - int64(strideDuration(timeInterval).Seconds())
+	}
+	firstCached, _, err := b.db.filter(&assetIdentity{source.ChainID(), address, source.Currency()}, nil, &balanceFilter{fetchTimestamp, maxAllRangeTimestamp, expandFlag(timeIntervalToBitsetFilter[timeInterval])}, 1, asc)
+	if err != nil {
+		return err
+	}
+
+	var oldestCached *big.Int
+	var oldestTimestamp int64
+	var newestCached *big.Int
+	if len(firstCached) > 0 {
+		oldestCached = new(big.Int).Set(firstCached[0].block)
+		oldestTimestamp = firstCached[0].timestamp
+		lastCached, _, err := b.db.filter(&assetIdentity{source.ChainID(), address, source.Currency()}, nil, &balanceFilter{minAllRangeTimestamp, maxAllRangeTimestamp, expandFlag(timeIntervalToBitsetFilter[timeInterval])}, 1, desc)
+		if err != nil {
+			return err
+		}
+		newestCached = new(big.Int).Set(lastCached[0].block)
+	} else {
+		var fetchBlock *big.Int
+		lastEntry, _, err := b.db.getLastEntryForChain(source.ChainID())
+		if err != nil {
+			return err
+		}
+		if lastEntry != nil {
+			fetchBlock = new(big.Int).Set(lastEntry.block)
+		}
+		mostRecentDataPoint, mostRecentBlock, err := b.fetchAndCache(ctx, source, address, fetchBlock, timeIntervalToBitsetFilter[timeInterval])
+		if err != nil {
+			return err
+		}
+
+		oldestCached = new(big.Int).Set(mostRecentBlock)
+		oldestTimestamp = int64(mostRecentDataPoint.Timestamp)
+		newestCached = new(big.Int).Set(mostRecentBlock)
+	}
+
+	if oldestTimestamp > startTimestamp {
```

Comment:
why do we need to cache backward if the cache values are already fullfilled?

Reply:
This is to cover the first fetch when we don't have a historical point to start from and cache forward. The previous implementation was finding the start point and only cache forward, but estimating a start point is difficult.